### PR TITLE
example(example):add custom mnist example

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description
1. add custom step example image
2. add @pass_context,  used for pass Context obj
image

3. expose dataloader, PPLResultStorage, pass_context from sdk
4. update examples with above changes

## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [ ] Client
- [ ] Python-SDK
- [x] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/consts/__init__.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 import pathlib
+import threading
 
+thread_local = threading.local()
```
Comment (tianweidut) [inline client/starwhale/consts/__init__.py:4]: use `threading.local()` directly?

if main thread called `consts.thread_local`, the other threads import the unexpected `thread_local` ?

File: client/starwhale/api/_impl/dataset/loader.py [hunk 2]
```diff
@@ -153,3 +155,36 @@ def get_data_loader(
     include_user_raw = summary.include_user_raw if summary else False
     _cls = UserRawDataLoader if include_user_raw else SWDSBinDataLoader
     return _cls(dataset_uri, start, end, logger or _logger)
+
+
+def data_loader(
```
Comment (tianweidut) [inline client/starwhale/api/_impl/dataset/loader.py:160]: use another function name?  `data_loader` and `get_data_loader` are confused.

File: client/starwhale/__init__.py [hunk 3]
```diff
@@ -23,14 +23,19 @@
...
 
 __all__ = [
     "__version__",
     "PipelineHandler",
     "multi_classification",
     "step",
+    "pass_context",
     "Context",
+    "Evaluation",
+    "get_data_loader_by_sharding",
```
Comment (tianweidut) [inline client/starwhale/__init__.py:38]: `get_data_loader_by_sharding` -> `get_sharding_data_loader` ?

File: client/starwhale/api/_impl/wrapper.py [hunk 4]
```diff
@@ -92,10 +107,16 @@ def log(self, table_name: str, **kwargs: Any) -> None:
...
         self._log(self._get_datastore_table_name(table_name), record)
 
-    def get_results(self) -> Iterator[Dict[str, Any]]:
-        return self._data_store.scan_tables(
+    def get_results(self, from_bytes: bool = False) -> Iterator[Dict[str, Any]]:
+        for data in self._data_store.scan_tables(
             [data_store.TableDesc(self._results_table_name)]
-        )
+        ):
+            if from_bytes:
```
[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/api/_impl/wrapper.py:114]: use another word to express `deserialize`?  `from_bytes` maybe confuse.

File: client/starwhale/consts/__init__.py [hunk 5]
```diff
@@ -1,5 +1,7 @@
 import pathlib
+import threading
 
+context_holder = threading.local()
```
Comment (tianweidut) [inline client/starwhale/consts/__init__.py:4]: Do we move `context_holder` to `starwhale.api._impl.job`?  `starwhale.consts` module stores the unchanged const.

File: example/mnist/mnist/custom_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+from .model import Net
+
+ROOTDIR = Path(__file__).parent.parent
+
+
+class CustomPipelineHandler:
+    def __init__(self) -> None:
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.model = self._load_model(self.device)
+        self.evaluation = t.Optional[None]
```
Comment (tianweidut) [inline example/mnist/mnist/custom_evaluator.py:31]: what's the `t.Optional[None]` meaning? 

`self.evaluation = t.Optional[None]` -> `self.evaluation = None`

File: example/mnist/mnist/custom_evaluator.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+    def __enter__(self) -> "CustomPipelineHandler":
```
Comment (tianweidut) [inline example/mnist/mnist/custom_evaluator.py:33]: Users should need to define `__enter__` and `__exit__` ? Is it required?

File: client/starwhale/__init__.py (hunk 3 above)
Comment (goldenxinxing) [reply client/starwhale/__init__.py]: done

File: client/starwhale/consts/__init__.py (hunk 5 above)
Comment (goldenxinxing) [reply client/starwhale/consts/__init__.py]: done

File: client/starwhale/api/_impl/wrapper.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply client/starwhale/api/_impl/wrapper.py:114]: > can all `from_bytes=True` results deserialize?
no, needs to be paired with ‘to_bytes=True’ on store to datastore

[the remark this answers is not in the document]
Comment (goldenxinxing) [reply client/starwhale/api/_impl/wrapper.py]: renamed

File: example/mnist/mnist/custom_evaluator.py (hunk 6 above)
Comment (goldenxinxing) [reply example/mnist/mnist/custom_evaluator.py]: already removed this unused property

File: example/mnist/mnist/custom_evaluator.py (hunk 7 above)
Comment (goldenxinxing) [reply example/mnist/mnist/custom_evaluator.py]: now give up using the 'with' syntax when instance an obeject

**Review by tianweidut — APPROVED** [review]
LGTM

**Pull request merged**